### DISCUSSION TITLE
Question about sample recording of AllocTracer

### DISCUSSION THREAD

**Discussion by yanglong1010** [async-profiler/async-profiler#961] [General]
Hi

The thread named `AllocThread` in the code below allocates 1 GiB memory.
If use Java 8 to run the code, the allocated memory reported by `JFR` is close to 1GiB.
but if use async-profiler with interval 10m to profile, the result is less than 500 MiB.

The reason is that we used `total_size` instead of `_interval`.
https://github.com/async-profiler/async-profiler/blob/09f0a8fef77a2c30ab0ffcd4841259cedb1be0ca/src/allocTracer.cpp#L51-L52

If the `_interval` is much larger than `total_size`, this will cause `async-profiler` to record fewer events, resulting in significantly lower overall memory allocation.

Can we modify it to use `_interval`, which would provide better accuracy ?

If you think it's okay, I'd like to contribute a PR.

```
java -version
openjdk version "1.8.0_412"
OpenJDK Runtime Environment (build 1.8.0_412-b08)
OpenJDK 64-Bit Server VM (build 25.412-b08, mixed mode)

java AllocTest
real total = 1024.0 MiB
jfr total = 934.7062225341797 MiB

java -agentpath:/root/async-profiler/build/lib/libasyncProfiler.so=start,alloc=10m,collapsed,total AllocTest|grep testAlloc
java/lang/Thread.run;AllocTest$$Lambda$52/1426329391.run;AllocTest.lambda$main$0;AllocTest.testAlloc;byte[]_[i] 431388168
```

```java
import jdk.jfr.Configuration;
import jdk.jfr.Recording;
import jdk.jfr.consumer.RecordedEvent;
import jdk.jfr.consumer.RecordingFile;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.concurrent.atomic.AtomicLong;

public class AllocTest {
    private static final AtomicLong n = new AtomicLong(0);

    public static void main(String[] args) throws Exception {
        Configuration recordingConfig = Configuration.getConfiguration("profile");
        try (Recording recording = new Recording(recordingConfig)) {
            recording.start();
            Thread t = new Thread(() -> {
                for (int j = 0; j < 1024; j++) {
                    for (int i = 0; i < 1024; i++) {
                        n.addAndGet(testAlloc(1024));
                    }
                }
            });
            t.setName("AllocThread");
            t.start();

            t.join();
            System.out.println("real total = " + (n.get() / 1024.0 / 1024) + " MiB");

            Path jfrPath = Files.createTempFile("jfr", ".jfr");
            recording.dump(jfrPath);
            try (RecordingFile file = new RecordingFile(jfrPath)) {
                long total = 0;
                while (file.hasMoreEvents()) {
                    RecordedEvent event = file.readEvent();
                    String name = event.getEventType().getName();
                    if ("jdk.ObjectAllocationInNewTLAB".equals(name)) {
                        if (event.getThread().getJavaName().equals("AllocThread")) {
                            total += event.getLong("tlabSize");
                        }
                    }
                    if ("jdk.ObjectAllocationOutsideTLAB".equals(name)) {
                        if (event.getThread().getJavaName().equals("AllocThread")) {
                            total += event.getLong("allocationSize");
                        }
                    }
                }
                System.out.println("jfr total = " + (total / 1024.0 / 1024) + " MiB");
            }
        }
    }

    private static int testAlloc(int size) {
        byte[] arr = new byte[size];
        return arr.length;
    }
}
```

**Comment by apangin** [comment]
For the comparsion with JFR to be fair, do not specify alloc interval:
```
-agentpath:libasyncProfiler.so=start,alloc,collapsed,total
```
In this case, `total` counter will also be about 930 MB.

Allocation profiling in JDK 8 relies on two different allocation events: `ObjectAllocationInNewTLAB` and `ObjectAllocationOutsideTLAB` which are not really comparable to each other. So, it's not trivial to create a fair sampling mechanism. If you have a good idea, I will be interested to know. `ObjectAllocationInNewTLAB` event has `tlabSize` field. At least, it currently has some physical meaning; it would be strange to record a user-specified interval in `tlabSize` field instead of a real TLAB size.

As I mentioned in #874, TLAB-based allocation events in JDK 8 are not good for making accurate estimations. They are still helpful for finding the hottest allocation sites. Due to rapidly decreasing popularity of JDK 8, I see no reason for investing in it much.

**Reply by yanglong1010** [reply]
Thanks for the reply. If I have good ideas about the sampling mechanism in the future, I'll open a new discussion. Regarding using `_interval` to replace `total_size`, I agree with your idea that this breaks the original meaning of `tlabSize`.
